From: Michal Vasko <mvasko@cesnet.cz>
Date: Fri, 1 Mar 2024 10:29:02 +0100
Subject: common BUGFIX do not use partial atoms

Causes redundant oper subscription calls.

(cherry picked from commit dd24ad78661f6564c1f00d8b9ff1aae7021eb5c2)
---
 src/common.c           |  4 ++--
 tests/test_oper_pull.c | 44 ++++++++++++++++++++++++++++++++++++++++++--
 2 files changed, 44 insertions(+), 4 deletions(-)

diff --git a/src/common.c b/src/common.c
index 6f1d951..c757e0f 100644
--- a/src/common.c
+++ b/src/common.c
@@ -5504,9 +5504,9 @@ parse_name:
     }
 
     if ((!end_chars && !next[0]) || (end_chars && next[0] && strchr(end_chars, next[0]))) {
-        /* finished with this (sub)expression, add new atom */
+        /* finished with this (sub)expression, add new atom if any found (parsed) */
         parsed = 1;
-        if ((err_info = sr_xpath_text_atom_add(&cur_atom, atoms, atom_count))) {
+        if ((strlen(prev_atom) < strlen(cur_atom)) && (err_info = sr_xpath_text_atom_add(&cur_atom, atoms, atom_count))) {
             goto cleanup;
         }
     } else {
diff --git a/tests/test_oper_pull.c b/tests/test_oper_pull.c
index fcffd11..a4e86bd 100644
--- a/tests/test_oper_pull.c
+++ b/tests/test_oper_pull.c
@@ -2075,7 +2075,9 @@ test_xpath_check(void **state)
     ret = sr_session_switch_ds(st->sess, SR_DS_OPERATIONAL);
     assert_int_equal(ret, SR_ERR_OK);
 
-    /* subscribe as state data provider */
+    /*
+     * subscribe as state data provider #1
+     */
     ret = sr_oper_get_subscribe(st->sess, "ietf-interfaces", "/ietf-interfaces:interfaces-state", xpath_check_oper_cb,
             st, 0, &subscr);
     assert_int_equal(ret, SR_ERR_OK);
@@ -2097,7 +2099,9 @@ test_xpath_check(void **state)
     sr_unsubscribe(subscr);
     subscr = NULL;
 
-    /* subscribe as state data provider */
+    /*
+     * subscribe as state data provider #2
+     */
     ret = sr_oper_get_subscribe(st->sess, "ietf-interfaces", "/ietf-interfaces:interfaces-state/interface[name='eth0']",
             xpath_check_oper_cb, st, 0, &subscr);
     assert_int_equal(ret, SR_ERR_OK);
@@ -2116,6 +2120,42 @@ test_xpath_check(void **state)
     sr_release_data(data);
     assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 1);
 
+    sr_unsubscribe(subscr);
+    subscr = NULL;
+
+    /*
+     * subscribe as state data provider #3
+     */
+    ret = sr_oper_get_subscribe(st->sess, "ietf-interfaces", "/ietf-interfaces:interfaces-state/interface/statistics",
+            xpath_check_oper_cb, st, 0, &subscr);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* set some operational data */
+    ret = sr_set_item_str(st->sess, "/ietf-interfaces:interfaces-state/interface[name='eth0']/type",
+            "iana-if-type:ethernetCsmacd", NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    ret = sr_set_item_str(st->sess, "/ietf-interfaces:interfaces-state/interface[name='eth0']/oper-status",
+            "down", NULL, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+    ret = sr_apply_changes(st->sess, 0);
+    assert_int_equal(ret, SR_ERR_OK);
+
+    /* read interfaces from operational, callback not called */
+    ATOMIC_STORE_RELAXED(st->cb_called, 0);
+    ret = sr_get_data(st->sess, "/ietf-interfaces:interfaces-state/"
+            "interface[not(derived-from-or-self(type, 'iana-if-type:softwareLoopback'))]/oper-status", 0, 0, 0, &data);
+    assert_int_equal(ret, SR_ERR_OK);
+    sr_release_data(data);
+    assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 0);
+
+    /* read all from operational, callback called */
+    ATOMIC_STORE_RELAXED(st->cb_called, 0);
+    ret = sr_get_data(st->sess, "/ietf-interfaces:interfaces-state/interface[name='eth0']", 0, 0, 0, &data);
+    assert_int_equal(ret, SR_ERR_OK);
+    sr_release_data(data);
+    assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 1);
+
     sr_unsubscribe(subscr);
 }
 
